### ISSUE TITLE
What was the reason to mark ImplicitStreamSubscriptionAttribute.Namespace property internal

### ISSUE THREAD

**Issue by centur** [dotnet/orleans#1266]
Hi, I have a question - what is the reason behind having Namespace property as internal in *ImplicitStreamSubscriptionAttribute*:

```
    [AttributeUsage(AttributeTargets.Class, AllowMultiple=true)]
    public sealed class ImplicitStreamSubscriptionAttribute : Attribute
    {
        internal string Namespace { get; private set; }

        // We have not yet come to an agreement whether the provider should be specified as well.
        public ImplicitStreamSubscriptionAttribute(string streamNamespace)
        {
            Namespace = streamNamespace;
        }
    } 
```

Namespace is used internally but streams semantic imply that user's code need to know namespace during subscription in OnActivateAsync, thus - namespaces must be hardcoded and can't be obtained from attribute itself using reflection over attribute. I want to create some kind of unified subscription mechanism in all grains for the following scenario:

Listener grain has multiple ImplicitStreamSubscription attributes, I want to extract all hardcoded Namespaces from them using reflection and mass-subscribe grain to all streams of events:

```[ImplicitStreamSubscription(nameof(CoreEventType.ProjectLocked))]
	[ImplicitStreamSubscription(nameof(CoreEventType.ProjectUnlocked))]
	[ImplicitStreamSubscription(nameof(CoreEventType.ProjectClosed))]
	[ImplicitStreamSubscription(nameof(CoreEventType.ProjectReopened))]
	[ImplicitStreamSubscription(nameof(CoreEventType.ProjectDeleted))]
	public class ProjectEventsListener: BaseProxyGrain, IGrainWithGuidKey, IStreamListener<CoreEventData>
	{
		private StreamSubscriptionHandle<CoreEventData>[] _handles;


		public override async Task OnActivateAsync()
		{
			var allImplicitAttributes = this.GetType()
				.GetCustomAttributes(typeof(ImplicitStreamSubscriptionAttribute), inherit: false);

			var fanoutTasks = allImplicitAttributes
				.Select(attrib => SubscribeToStream(StreamNames.CoreEvents_Provider, ((ImplicitStreamSubscriptionAttribute)attrib).Namespace))
				.ToList();

			await Task.WhenAll(fanoutTasks);

			_handles = fanoutTasks.Select(t => t.Result).ToArray();

			await base.OnActivateAsync();
		}
}
```

This is currently impossible without abusive reflection and peeking into each attribute instance's internal field. The approach I don't want to take. So can we make Namespace property publicly available ?

Also I've noted the comment on ```ImplicitStreamSubscriptionAttribute ctor```
> // We have not yet come to an agreement whether the provider should be specified as well.

Is there a discussion with arguments against adding StreamProvider param to attribute ctor ?

From our current usage I can add strong argument for adding extra ctor with 2 arguments - StreamProvider and Namespace, and make them publicly available as attribute properties. Then the code above can be implemented without hardcoding provider name. Even more - implicit subscription wiring can be moved into base class if the grain implements certain interface, in my example - ```IStreamListener``` is defined like this:

```	
public interface IStreamListener<TMessageType>
	{
		Task ReceiveCommand(TMessageType item, StreamSequenceToken token = null);
		Task OnCommandCompleted();
		Task OnCommandError(Exception ex);
	}
```

which allows us to subscribe grains automatically in certain cases.

**Comment by gabikliot** [comment]
>So can we make Namespace property publicly available

No objections. I don't think anyone will object. Just submit a PR.

As for the 2nd question re adding stream provider: I will let @jason-bragg answer. I have no objections, but he might have a different view.

**Comment by centur** [comment]
Thanks.

on 2nd> We can keep 2 ctors - with and without StreamProvider. So old code can still work with old semantic, new code can store some extra information - given that it should be compile time constant, it doesn't make much difference where to store it - in attribute instance or in static field, but storing it in attribute makes subscription semantic much easier and gives extra flexibility to Listeners - particular grain can be a listener to multiple stream providers using declarative semantic, without code to match namespace with specific stream provider name.

Side note: I'm also curios whether it's easy\possible to retrieve all namespaces from Streams infrastructure - I think I have couple ideas on how to implement wildcard based subscriptions for namespaces, similar to @yevhen's Orleanakka, which looks pretty straightforward  (effectively it'll be implicit subscriptions but user can use wildcards in Namespace param

**Comment by gabikliot** [comment]
What do you mean by: "retrieve all namespaces from Streams infrastructure"? What does "all" even mean? When/where all?

Perhaps you meant: would it be possible to provide a stream namespace with a wildcard at subscription time - that is, subscribe to all streams that start with the given prefix?
And how Orleanakka does that?

**Comment by centur** [comment]
Disregard the side note - while trying to explain my concept here, I figured out that the idea behind it is not possible, even if I can get all statically defined namespaces (via reflection) - it's not going to work the way I thought it's working internally. Although it's possible to subscribe to all static namespaces via that concept - in runtime client can push a message to a stream with random namespace - thus we can't catch this message without explicit subscription unless we change something in PubSub implementation (this is an educated guess).

Orleankka (from my shallow studying ) has it's own implementation ```class StreamSubscriptionMatcher : IStreamProviderImpl``` which is doing all the magic. But I suspect that it's just @yevhen own implementation which returns custom Stream which sends  msg to all matched recipients here

```
  public IAsyncStream<T> GetStream<T>(Guid unused, string id)
        {
            var stream = provider.GetStream<T>(unused, id);

            return (IAsyncStream<T>) streams.GetValue(stream, _ =>
            {
                var recipients = Match(system, id, specifications);

                Func<T, Task> fan = item => TaskDone.Done;

                if (recipients.Length > 0)
                    fan = item => Task.WhenAll(recipients.Select(x => x.Receive(item)));

                return new Stream<T>(stream, fan);
            });
        }
```

**Comment by jason-bragg** [comment]
@centur 
Question about your scenario:
Do you expect to have several streams processed by the same grain, for instance the following streams.
streamGuid1, CoreEventType.ProjectUnlocked
streamGuid1, CoreEventType.ProjectClosed
streamGuid1, CoreEventType.ProjectReopened
streamGuid1, CoreEventType.ProjectDeleted
Or are you processing a single stream that may be one of the 4 namespaces above, and you want to subscribe to all 4 namespaces to ensure you subscribe to the active stream?

**Comment by centur** [comment]
I think we will ended up in both worlds:

Right now we have first scenario - one grain, which is responsible for handling all events around project changes. But it already getting too much responsibility: it's doing calls to other grains to perform updates and it also has duty to notify users. So we are going to move some responsibilities to other grains. 

Current plans are to have many grains subscribed to their areas of interest - e.g data-updating grain is subscribed to ProjectDeleted\ProjectCreated and other Project* namespaces, because it's responsible for updating other participants of the "graph", but Notification grain is subscribed to all user\project\global events and just sends updates to all interested users.

**Comment by jason-bragg** [comment]
@centur 
Thanks for the details.  The issue you've raised in this thread is one we've (internally) discussed before, but have not yet addressed.  In general, the existing implicit subscription system has no means of providing the fully qualified streamId (streamGuid, streamNamespace, streamProviderName) of the activating stream to the grain.  I'm hesitant to just make the attribute's stream namespace (and/or provider name) public as that does not really solve the problem, as it only allows the grain to query which namespaces/providers it 'could' receive messages on, not which(s) one it was activated to service.

The workaround we've used in the past for grains that could subscribe to a single stream, but from multiple namespaces or stream providers, is to create an abstract base class with StreamProviderName and StreamNamespace properties that it uses to subscribe to the stream, then define a subclass of it for each namespace/streamprovider combination it could subscribe to.  This is tedious, and should not be necessary.

While I'm not adamantly opposed to what is proposed in this thread, I consider it a band-aid that masks the underlying issue.  I would prefer a proposal that introduces some way for the activated grain to get access to the fully qualified streamId of the activating stream.  I suspect such a solution would require a base class, which has been mentioned, which I don't oppose.

**Comment by centur** [comment]
I support the opinion that Streaming system needs improvements, peeking into source code makes me feel that it's designed with very complex architecture in mind. 
Still with the current implementation we can make some minor updates that allow us to make usage of it a bit more  easier. Yes it's band-aiding but sometimes you need to band-aid nasty issues first before you have time to re-implement entire subsystem.

**Comment by gabikliot** [comment]
@centur: Re wildcard subscription:
assuming that what you meant above is "would it be possible to provide a stream namespace with a wildcard at subscription time - that is, subscribe to all streams that start with the given prefix" and NOT  "retrieve all namespaces from Streams infrastructure" (which I don't know what does "all" even mean), my answer is:

Yes, it is definitely possible. Yes, it would requite a (small) change in the pub sub, which is easy. It will be easy for implicit subscriptions. We will just change this line: https://github.com/dotnet/orleans/blob/master/src/Orleans/Streams/PubSub/ImplicitStreamSubscriberTable.cs#L62 to do wildcard comparison.

Re explicit subscriptions: that would be harder, as we now have [one pubsub grain per stream Guid + stream namespace](https://github.com/dotnet/orleans/blob/master/src/Orleans/Streams/PubSub/GrainBasedPubSubRuntime.cs#L61), so wildcard  would not work out of the box. One way to solve it will be by adding another optional pub sub where RendezvousGrains are identified only by steam Guid, and demultiplexing inside. More work, but doable.

The bigger issue with wildcard subscription is what  @jason-bragg pointed to:
"the existing implicit subscription system has no means of providing the fully qualified streamId (streamGuid, streamNamespace, streamProviderName) of the activating stream to the grain"
This also holds for explicit.

That mean that you will NOT be able to differentiate in the OnNext what stream the msg comes on.

As opposite to PR #1270 (which I think we should merge), here I agree more with @jason-bragg . Although we **could** allow wildcard subscription and make it work in the pub sub, the bigger question is whether we **should**. If the fact that one cannot differentiate the source stream makes this feature not usable, then we should NOT add it. If we get convinced that it is still usable (albeit in a more limited context), that we could consider adding it.

**Comment by centur** [comment]
@jason-bragg: thoroughly re-read your answer and realized that I need more knowledge about internal implementation before I can comprehend the answer :(. Is the following recaps what you've said there:
"There is no way for the grain to know which stream activated it because the activation can be caused by a message for one FullyQualifiedStream (Id-Provider-Namespace)  but grain may subscribe\register for completely different FQStream". Is this what you're saying in the first paragraph? I've realized that attribute is something like a flag to trigger some pre-activation, rather than a strict way to limit the listener, is this right ?

@gabikliot : just ignore "all namespaces", later on I realized that it's impossible to have "all", the only list I may get in that context is "all namespaces from attributes" (which are compile-time), but this has no use  because producer can produce events with any random namespace.

After your comment I think I understand your's and Jason's concerns around wildcards. 
Re differentiating what stream the message comes on - why is this important ? For me a stream concept is a like a fire hose or "river of events" - do we need to differentiate the source for every drop's in a water stream or we just want to process each drop, no matter where it's coming from?

From stream identification perspective I may use some RequestContext data, but I understand that other systems mileage may vary.

**Comment by jason-bragg** [comment]
@centur 
The ImplicitSubscription attribute tells the streaming infrastructure to send events to an instance of this grain for any stream on any provider using this namespace.  GrainId will be the same as the streamGuid.  When the streaming infrastructure sends events to an instance of this grain (with grain ID == streamGuid) the grain is woken up, and must subscribe to the stream to assign an observer to process the events.  The issue here is that the grain only knows the streamGuid (because it's the same as the grainId) but does not know the stream provider or namespace of the stream which activated the grain.

In most cases systems have only a single stream provider and implicit subscription grains are attributed with only a single namespace, so the StreamProviderName and StreamNamespace can be hard coded into the grain, but this is quite limiting for scenarios like yours.

Now imagine that stream grains inherited from a base class that had an overridable call something like
  virtual Task OnNewStream(Guid streamGuid, string streamNamespace, string streamProviderName)
which was called prior to sending the first event to on as stream to a grain.  The OnActivateAsync would not need to subscribe, instead subscribing to the stream would be done in the OnNewStream call using all the information needed to subscribe to the exact stream that is sending events to it.  If the grain could receive events from 20 namespaces on 20 stream providers (total of 400 possible combinations) instead of subscribing to all of these on activate, it would subscribe only to the active stream as they become active.  I'm not suggesting this as a solution, just as an example to help illustrate the missing capability.

**Comment by gabikliot** [comment]
Fixed via #1270.
